ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPrivate.qll

```diff
@@ -1972,7 +1972,26 @@ private predicate mustHaveCollectionType(Node n, DataFlowType t) {
   not n instanceof SynthSplatParameterNode
 }
 
-predicate localMustFlowStep(Node node1, Node node2) { none() }
+predicate localMustFlowStep(Node node1, Node node2) {
+  node1 = SsaFlow::toParameterNodeImpl(node2)
+  or
+  exists(SsaImpl::Definition def |
+    def.(Ssa::WriteDefinition).assigns(node1.asExpr()) and
+    node2.(SsaDefinitionExtNode).getDefinitionExt() = def
+    or
+    def = node1.(SsaDefinitionExtNode).getDefinitionExt() and
+    node2.asExpr() = SsaImpl::getARead(def)
+  )
+  or
+  node1.asExpr() = node2.asExpr().(CfgNodes::ExprNodes::AssignExprCfgNode).getRhs()
```

Comment:
This looks good to me, I was just wondering if there was overlap between this line and `def.(Ssa::WriteDefinition).assigns(node1.asExpr())` above. Both seem to be related to assignments. Similar question about `SsaFlow::toParameterNodeImpl` and the special handling for `(ImplicitBlockArgumentNode).getParameterNode(_)`.

Reply:
> This looks good to me, I was just wondering if there was overlap between this line and def.(Ssa::WriteDefinition).assigns(node1.asExpr()) above. Both seem to be related to assignments.

The first line is about flow from `e` to SSA def for `x` in `x = e`, and the second line is about flow from `e` to the whole assignment expression `x = e`.

> Similar question about SsaFlow::toParameterNodeImpl and the special handling for (ImplicitBlockArgumentNode).getParameterNode(_).

This is because implicit block parameters and arguments are not covered by our SSA analysis, otherwise you are right it would be an instance of that.

Reply:
Thanks for clarifying.